docs/architecture-decisions/2023-01-11_beaker_productionisation_review.md

```diff
@@ -0,0 +1,409 @@
+
+This change will simplify Beaker's code (improving maintainability) and, more importantly, reduce the potential for end-user error and confusion.
+
+#### Why?
+
+**User-facing benefits**
+
+1. The current structure, by encouraging and supporting [bound instance methods](https://www.geeksforgeeks.org/bound-methods-python/) alongside [class variables](https://pynative.com/python-class-variables/), is a potential source of confusion for users new to writing smart contracts or PyTEAL. The distinction between what runs on `beaker.Application` instantiation, evaluation by PyTeal during compile, and finally what runs on-chain, can be difficult to grasp at first. One might assume (wrongly) that Beaker is somehow maintaining the state of `self.*` between methods, but this is not the case. Contrast this with Solidity, for example, where state can be directly manipulated because it's help within the class instance.
+2. Currently, actually using `self.*` can easily lead to problems, since if they are not defined before calling `super().__init__(...)` they won't be defined when compiling. This can be fixed by not automatically compiling in `Application.__init__()` (which is also proposed in (2) below) for simple constants, however another issue is that using `self.foo = <Some beaker.state object>`, would not currently work with the introspection beaker is performing. This could potentially be fixed by itself, but developers will still need to define these values _before_ calling `super().__init__()` which is a source of confusion. Usually, idiomatic Python will call super init sooner rather than later so this is something that can trip up experienced Python developers.
+3. In order to compose applications together, say if there were two ARC standard implementations that we wanted to combine into the same contract, the user doesn't need to understand Python's multiple-inheritance idiosyncrasies like [Method Resolution Order](https://www.geeksforgeeks.org/method-resolution-order-in-python-inheritance/). Additionally, by taking a functional composition approach, we can have easy to understand entry points where you can check any pre-conditions.
+4. Since state variables are currently defined as class variables, this makes them "globals", which can lead to errors/bugs that are non-obvious.
+
+   For instance, consider:
+
+   ```python
+   class MyBaseApp(beaker.Application):
+       counter = beaker.ApplicationStateValue(stack_type=pyteal.TealType.uint64)
+
+       @beaker.create
+       def create(self) -> pyteal.Expr:
+           return self.initialize_application_state()
+
+   class MyApp(MyBaseApp):
+       pass
+   MyApp.counter.default = pyteal.Int(10)
+
+   class MyOtherApp(MyBaseApp):
+       pass
+
+   app1 = MyApp()
+   app2 = MyOtherApp()
```

Comment:
I considered it and now my brain hurts.  It seems more like a python idiosyncrasy but it does illustrate your point.